### ISSUE TITLE
MeiliSearchEngine does not work with custom Scout keys

### ISSUE THREAD

**Issue by DarronEngelbrechtEdge** [laravel/scout#653]
- Scout Version: 9.4.10
- Scout Driver: MeiliSearch
- Laravel Version: 9.21.6
- PHP Version: 8.1.9
- Database Driver & Version: MySQL 8.0
- SDK Version (If using a third-party service): N/A
- Meilisearch CLI Version (If using Meilisearch): meilisearch-http 0.28.0

### Description:

When using custom index keys with `getScoutKey()` in a Searchable model the MeiliSearchEngine does not correct add the custom key to index nor delete entries from the index.

1. The `update()` method in the MeiliSearchEngine class uses array_merge() to merge the Scout key index with the toSearchableArray() data but references the Scout key first so the custom key gets overwritten by $searchableData unless a custom key is used.

```php
           return array_merge(
               // [$model->getKeyName() => $model->getScoutKey()],
                $searchableData,
                $model->scoutMetadata(),
                [$model->getKeyName() => $model->getScoutKey()]
            );
```

2. RemoveFromSearch job makes a RemoveableScoutCollection which runs the models through `getScoutKey()` and passes the result to the the `delete()` method in MeiliSearchEngine which then runs `getScountKey()` a second time.
ie.
```php
    /**
     * Remove the given model from the index.
     *
     * @param  \Illuminate\Database\Eloquent\Collection  $models
     * @return void
     */
    public function delete($models)
    {
        $index = $this->meilisearch->index($models->first()->searchableAs());

        $index->deleteDocuments(
            // $models->map->getScoutKey()
            //     ->values()
            //     ->all()
            $models->pluck($models->first()->getKeyName())->toArray()
        );
    }
```
PS. We may also want to look at replacing getKeyName() with getScoutKeyName() in this MeiliSearchEngine class?

### Steps To Reproduce:

1. Create a searchable model which overrides the `getScoutKey` method to prepend a something...
```php
    /**
     * Get the custom search index key
     *
     * @return void
     */
    public function getScoutKey()
    {
        return 1 . $this->getKey();
    }
```
3. Create and `->save()` the model, notice in Meilisearch the ID is not prepended with 1.

4. Delete the model, notice the model is not removed from Meilisearch.

**Comment by mmachatschek** [comment]
@DarronEngelbrechtEdge if you want to override the value of the primary key in meilisearch you would need to override the `->toSearchableArray()` method too.

Unfortunately the documentation is not clear enough in this case.

By default the `toSearchableArray()` method just creates an array form of the model and sends the payload to the underlying search engine.

@driesvints no duplicate of #651 as this is about the indexing. Though after fixing the searchable array on this instance @DarronEngelbrechtEdge will run into the issue as described in #651

**Comment by DarronEngelbrechtEdge** [comment]
@mmachatschek and @driesvints 

Thanks for the feedback, @mmachatschek is 100% correct regarding point 1, the following works to load the custom Scout key in the index...

```
    public function toSearchableArray()
    {
        return [
            'id' => $this->getScoutKey()
        ];
    }
```
Maybe this issue can be fixed with documentation rather than switching the order in the array_merge() as I suggested in point 1 of the issue description which will just change the behavior and may still seem strange to some. 


Regarding point 2 here, 

Once you manage to get the custom scout key into the index the RemoveFromSearch job does not actually remove the model from the index due to sending in the incorrect ID, it sends in the incorrect ID because...

`RemoveFromSearch` uses `RemoveableScoutCollection` to build a collection using `->getScoutKeys()` to return the correct IDs and then runs the `delete()` method on the configured engine passing in the RemoveableScoutCollection. All 100% up to here unless there are issues in other engines that I'm unaware of.... 

The `delete()` method in the MeiliSearchEngine however tries to map the collection through ->getScoutKeys() again which it does not need to do but could perhaps just pluck out the IDs from the collection?

**Comment by francoism90** [comment]
@DarronEngelbrechtEdge 
I'm having the issue, using your workaround indeed fixes the issue:
```
public function toSearchableArray(): array
    {
        return [
            'id' => $this->getScoutKey(), // this one
            'name' => $this->name,
            'email' => $this->email,
        ];
    }
```

**Comment by DarronEngelbrechtEdge** [comment]
@mmachatschek @driesvints - I have opened a draft PR #655  to demonstrate the issue with getScoutKey being calculated from a field by hashing the value in SearchableModelWithCustomKey. I'm not too sure how to link it to this issue. Hope this helps to understand the issue, will see if I can take a stab at fixing it a bit later today. Many thanks!

**Comment by DarronEngelbrechtEdge** [comment]
@mmachatschek @driesvints - I tried to replicate my issue directly using the unit test below but I was getting different results in the Scout package (SCOUT_QUEUE=false) as to what I get in my project with SCOUT_QUEUE=true. Switching SCOUT_QUEUE=false in the project gives me the same results as in the unit test (things would work 100%).

Thus, I believe my PR demonstrates the core issue as the unit test `test_models_are_deserialized_without_the_database_using_custom_scout_key` fails due to running through `getScoutKey()` twice?


```php
    public function test_delete_removes_objects_to_index_using_custom_scout_key()
    {
        Config::shouldReceive('get')->with('scout.prefix')->andReturn('');

        $client = m::mock(Client::class);
        $client->shouldReceive('index')->with('table')->andReturn($index = m::mock(Indexes::class));
        $index->shouldReceive('deleteDocuments')->with(['my-meilisearch-key.1']);

        $engine = new MeiliSearchEngine($client);
        $engine->delete(RemoveableScoutCollection::make([new MeiliSearchCustomKeySearchableModel(['id' => 1])]));
    }
```

**Comment by DarronEngelbrechtEdge** [comment]
For anybody waiting on the outcome here I have worked around this overriding the RemoveFromSearch job for the application using the changes in PR #655 - to override the job just add something like the following to your AppServiceProvider.php - thanks Laravel and Scout - that was easy! :-)

```php
 Scout::removeFromSearchUsing(\App\Jobs\Scout\RemoveFromSearch::class);
```
